uiux/challenge-1/src/components/tracks.tsx

```diff
@@ -0,0 +1,105 @@
+import * as React from 'react';
+
+import Track from '../models/track';
+
+const ASC = 'asc';
+const DESC = 'desc';
+
+const DEFAULT_SORT = {
+    name: null,
+    length: null
+};
+
+export default class Tracks extends React.Component<{ sorted: Track[],tracks:Track[] }> {
+    sort: object;
+
+    constructor(props) {
+        super(props);
+
+        this.sort = {...DEFAULT_SORT};
+
+        this.state = {
+            lyrics: null
+        };
+    }
+
+    getSortCaret(column) {
+        let up = 'fas fa-caret-up', down = 'fas fa-caret-down';
+        if (this.sort[column] !== null) {
+            if (this.sort[column] === ASC) {
+                down += ' white'
+            } else {
+                up += ' white';
+            }
+        }
+        return <div className='carets'>
+            <i className={up}></i>
+            <i className={down}></i>
+        </div>;
+    }
+
+    showLyrics(track) {
+        track.getLyric().then(lyrics => {
+            const l = lyrics.replace(/[\r\n]/g, '<br>');
+            this.setState({ lyrics: l });
+        })
+    }
+
+    sortBy(column) {
+        const goAsc = this.sort[column] === DESC || this.sort[column] === null;
+        this.sort = { ...DEFAULT_SORT };
+        this.sort[column] = !goAsc ? DESC : ASC;
+        this.props['tracks'].sort((a, b) => {
+            if (goAsc) {
+                if (a[column] > b[column]) return 1;
+                if (a[column] < b[column]) return -1;
+                return 0;
+            } else {
+                if (a[column] < b[column]) return 1;
+                if (a[column] > b[column]) return -1;
+                return 0;
+            }
+        });
+        this.setState({});
```

Comment:
Well, dealing with `props` and `State` can get a bit messy. So to trigger a `render` on the page without calling the actual `render` method, which doesn't work anyway... Calling the `setState` with an empty object does the trick. Not the cleanest, but it will do for now...